### PR TITLE
[libc++] Implement LWG3940: std::expected<void, E>::value() also needs E to be copy constructible

### PR DESCRIPTION
**Pull request by PragmaTwice**
This patch includes the fix for [LWG3940](https://cplusplus.github.io/LWG/issue3940) (`std::expected<void, E>::value()` also needs `E` to be copy constructible)

### CODE REVIEW

File: libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.compile.fail.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+//===----------------------------------------------------------------------===//
```
Comment (huixie90) [inline libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.compile.fail.cpp:1]: We'd like to avoid writing compile.fail.cpp. Instead, verify.cpp is better because it can check the exact line that causes the compilation error. an example can be found here:

https://github.com/llvm/llvm-project/blob/main/libcxx/test/libcxx/utilities/expected/expected.unexpected/swap.mandates.verify.cpp

Comment (PragmaTwice) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.compile.fail.cpp]: Great idea! Thank you.

Done. Could you review it again when you have any free time?

File: libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+void test() {
+  // MoveOnly type as error_type
+  std::expected<void, MoveOnly> e(std::unexpect, 5);
+
+  std::move(e)
+      .value(); // expected-note{{in instantiation of member function 'std::expected<void, MoveOnly>::value' requested here}}
+  // expected-error-re@*:* {{static assertion failed due to requirement 'is_copy_constructible_v<MoveOnly>': error_type has to be both copy constructible and move constructible}}
+}
+
+int main(int, char**) { test(); }
```
Comment (huixie90) [inline libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp:27]: we don't need this line. the verify test is not meant to be run

File: libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+//===----------------------------------------------------------------------===//
```
Comment (huixie90) [inline libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp:1]: let's move this test under libcxx/test/libcxx/[utilities/expected/expected.void
The folder "std" is for tests that should pass on all implementations (not only libc++, and AFAIK, MSVC STL uses our tests under libcxx/test/std)

File: libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp [hunk 4]
```diff
@@ -65,17 +65,6 @@ void testException() {
     }
   }
 
-  // MoveOnly
```
Comment (huixie90) [inline libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp:68]: Out of interest, according to the LWG issue, this test should never work. Do you know why there was not an issue before?

**Review by huixie90 — COMMENTED** [review]
LGTM with all issues are addressed. Thank you!

File: libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+// ADDITIONAL_COMPILE_FLAGS: -Xclang -verify-ignore-unexpected=error
```
Comment (mordante) [inline libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp:9]: Why did you add this line?

**Review by mordante — COMMENTED** [review]
Thanks for your patch!

File: libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp (hunk 5 above)
Comment (PragmaTwice) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.lwg3940.verify.cpp]: Sorry I just copy from example file in https://github.com/llvm/llvm-project/pull/71819#discussion_r1394171789, and forget to delete this flag.

Now it's deleted.

File: libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp (hunk 4 above)
Comment (PragmaTwice) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp:68]: Yeah due to LWG3940 it cannot pass the compiler. But I guess before this the `expected::value` implementation can actually accept move-only types.

**Comment by PragmaTwice** [comment]
It seems the CI has failed, but I'm not sure if it is related to my changes, since I don't touch these failed test cases.

```
std/containers/container.adaptors/container.adaptors.format/format.functions.format.pass.cpp' FAILED
```

**Comment by PragmaTwice** [comment]
There are still some random CI errors, but seems not related to my changes.

Hi @huixie90 @mordante, could you review this PR again when you are free?

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+// constexpr void value() &&;
+// Mandates: is_copy_constructible_v<E> is true and is_move_constructible_v<E> is true.
```
Comment (mordante) [inline libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:11]: Can you add tests for all mandates?

**Review by huixie90 — APPROVED** [review]
LGTM

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <expected>
+
+#include "MoveOnly.h"
+
+void test() {
```
Comment (huixie90) [inline libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:17]: Could you add a test for none `void` case too?

**Review by mordante — CHANGES_REQUESTED** [review]
@PragmaTwice There are some review comments still open. Do you have time to address these? Then we can merge this patch.

**Comment by PragmaTwice** [comment]
> @PragmaTwice There are some review comments still open. Do you have time to address these? Then we can merge this patch.

Apologies for my delayed response. I intend to complete it this week.

File: libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp (hunk 4 above)
Comment (frederick-vs-ja) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp:68]: > Out of interest, according to the LWG issue, this test should never work. Do you know why there was not an issue before?

This is because of Clang's non-conforming acception of move-only exception types, which is reported as #53224.

Comment (philnik777) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp:68]: Why would this extension be non-conforming? A warning is missing, but other than that I don't see how it wouldn't be conforming.

Comment (frederick-vs-ja) [reply libcxx/test/std/utilities/expected/expected.void/observers/value.pass.cpp:68]: It seems that `throw NonCopyable(args)` also needs result in substitution failure, while Clang currently disagrees ([example](https://godbolt.org/z/x1j16E54K)). See also [this article](https://quuxplusone.github.io/blog/2021/03/18/sfinae-on-throw-x/).

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp (hunk 7 above)
Comment (PragmaTwice) [reply libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:26]: Done : )

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp (hunk 6 above)
Comment (PragmaTwice) [reply libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:14]: Done : )

File: libcxx/include/__expected/expected.h [hunk 8]
```diff
@@ -1187,12 +1187,15 @@ class expected<_Tp, _Err> {
   }
 
   _LIBCPP_HIDE_FROM_ABI constexpr void value() const& {
+    static_assert(is_copy_constructible_v<_Err>, "error_type has to be copy constructible");
```
[the remark this answers is not in the document]
Comment (PragmaTwice) [reply libcxx/include/__expected/expected.h]: Done.

**Comment by PragmaTwice** [comment]
Hi @mordante @huixie90, I think all review comments have been addressed. Could you review it again?

**Comment by huixie90** [comment]
> Hi @mordante @huixie90, I think all review comments have been addressed now. Could you review it again?

Hi thanks for the update. There 2 expected tests in the CI  that seem to be failing. Do you know why it is the case?

**Comment by PragmaTwice** [comment]
> > Hi @mordante @huixie90, I think all review comments have been addressed now. Could you review it again?
> 
> Hi thanks for the update. There 2 expected tests in the CI that seem to be failing. Do you know why it is the case?

Yeah I've seen these failed cases.
But I cannot reproduce on my side, and it seems not related to my changes.

I will try to merge the latest main branch and rerun the CI.

```
# | error: 'expected-error' diagnostics seen but not expected: 
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1378: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1378: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1389: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1389: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1400: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1400: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1412: excess elements in struct initializer
# |   File /home/runner/_work/llvm-project/llvm-project/build/generic-modules/include/c++/v1/__expected/expected.h Line 1412: excess elements in struct initializer
```

**Comment by PragmaTwice** [comment]
https://github.com/llvm/llvm-project/actions/runs/7404652886/job/20147039107?pr=71819
```
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen: 
# |   File * Line * (directive at /home/runner/_work/llvm-project/llvm-project/libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:46): call to deleted constructor of 'MoveOnly'
# |   File * Line * (directive at /home/runner/_work/llvm-project/llvm-project/libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:47): call to deleted constructor of 'CopyOnly'
# |   File * Line * (directive at /home/runner/_work/llvm-project/llvm-project/libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:48): call to deleted constructor of 'CopyOnly'
# | 3 errors generated.
# `-----------------------------
```

It seems in some specific version/config of clang, these errors do not appear.

May I ask if there is any method to workaround these non-important error messages?

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+  e2.value();
+
+  std::move(e2)
+      .value(); // expected-note {{in instantiation of member function 'std::expected<void, CopyOnly>::value' requested here}}
+  // expected-error@*:* {{static assertion failed due to requirement 'is_move_constructible_v<CopyOnly>': error_type has to be both copy constructible and move constructible}}
+
+  // expected-error@*:* {{call to deleted constructor of 'MoveOnly'}}
+  // expected-error@*:* {{call to deleted constructor of 'CopyOnly'}}
+  // expected-error@*:* {{call to deleted constructor of 'CopyOnly'}}
```
Comment (huixie90) [inline libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:48]: The CI is complaining about these lines when the test is running with no-exceptions mode or no-rtti mode.
perhaps just mark the test unsupported for these settings?

```
// UNSUPPORTED: no-rtti
// UNSUPPORTED: no-exceptions
```

Comment (PragmaTwice) [reply libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:50]: Ahhh I see. Thank you!

Comment (PragmaTwice) [reply libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:50]: Fixed : )

**Comment by PragmaTwice** [comment]
The CI finally passed now : )

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (mordante) [inline libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp:50]: Can you move these expected-error's to the places where they are generated. We need the `@*:*` since they are in the headers, but for maintainability it's good to have them where they are generated.

**Review by mordante — COMMENTED** [review]
> The CI finally passed now : )

Great. A few minor comment and then it LGTM. I'll leave the final approval to @huixie90

File: libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp (hunk 10 above)
Comment (PragmaTwice) [reply libcxx/test/libcxx/utilities/expected/expected.void/value.lwg3940.verify.cpp]: done in 2614080a0e5260021a80140e5096bb86b46ebf23

File: libcxx/include/__expected/expected.h [hunk 11]
```diff
@@ -1150,12 +1150,15 @@ class expected<_Tp, _Err> {
...
   _LIBCPP_HIDE_FROM_ABI constexpr void value() const& {
+    static_assert(is_copy_constructible_v<_Err>);
     if (!__has_val_) {
       std::__throw_bad_expected_access<_Err>(__union_.__unex_);
     }
   }
 
   _LIBCPP_HIDE_FROM_ABI constexpr void value() && {
+    static_assert(is_copy_constructible_v<_Err> && is_move_constructible_v<_Err>,
+                  "error_type has to be both copy constructible and move constructible");
```
[the remark this answers is not in the document]
Comment (PragmaTwice) [reply libcxx/include/__expected/expected.h]: done in 2614080a0e5260021a80140e5096bb86b46ebf23

**Comment by PragmaTwice** [comment]
> > The CI finally passed now : )
> 
> Great. A few minor comment and then it LGTM. I'll leave the final approval to @huixie90

Thank you! Comments has been addressed now.

**Review by mordante — APPROVED** [review]
> LGTM. @mordante could you please have another look if the current state addressed your concerns?

Yes looks fine to me.
